### PR TITLE
Avoid resetting the session key with each call to put/4

### PR DESCRIPTION
**Pull request by foxbenjaminfox**
After upgrading to the new v0.7.0 of redbird, I found I consistently got `Plug.Conn.CookieOverflowError`s after a few requests.

This is strange—the session key is supposed to be stable, yet it was changing and growing in size with each request.

When I examined the overly-large value being set in the cookie, I saw that it was something signed with `Plug.Crypto.sign`, whose payload is a value _also_ signed by `Plug.Crypto.sign`, and so forth, quite a few levels deep.

In other words: The redis key gets re-signed by `Redbird.Crypto.sign` with every call to `put/4`—even if that key _was already_ signed, leading to a new (larger) key with each request!

The fix is to sign the key exactly once, right after generating it, then verify that the key is signed properly in `put/4` in the same way as in `get/3` or `delete/2`.

### CODE REVIEW

**Comment by germsvel** [comment]
@foxbenjaminfox thanks so much for opening this PR! I'll give it a look as soon as I can. 

@lancejjohnson I know you worked on this, so I also wanted to reach out in case you could give it a look.

File: lib/redbird/plug/session/redis.ex [hunk 1]
```diff
@@ -25,17 +24,24 @@ defmodule Plug.Session.REDIS do
   end
 
   def put(conn, nil, data, init_options) do
-    put(conn, Key.generate(), data, init_options)
+    key =
+      Key.generate()
+      |> Key.sign_key(conn)
```
Comment (germsvel) [inline lib/redbird/plug/session/redis.ex:29]: Ah, yes. This is a good catch 👍

I think this is where we're supposed to sign the key since it's the first time it's generated.

File: lib/redbird/plug/session/redis.ex [hunk 2]
```diff
@@ -25,17 +24,24 @@ defmodule Plug.Session.REDIS do
...
+
+    put(conn, key, data, init_options)
   end
 
   def put(conn, key, data, init_options) do
-    key
-    |> Key.sign_key(conn)
-    |> set_key_with_retries(Value.serialize(data), session_expiration(init_options), 1)
+    if Key.accessible?(key, conn) do
```
Comment (germsvel) [inline lib/redbird/plug/session/redis.ex:35]: Do you know if we need to make this check here? 

Looking at [Plug.Session.Store's docs](https://hexdocs.pm/plug/Plug.Session.Store.html#c:put/4), I see that we're supposed to generate the key when its value is `nil` (which is what you're doing in the function above). But I think after that we can just take the `key` that Plug is giving us and setting it.

At the very least, I don't think we should be returning the `key` in the `else` clause below since it hasn't passed the `Key.accessible?` check. In that case, we're not actually setting the key, but by returning it, we're incorrectly telling Plug that we've storing the session. 

So, do you know if there's harm in just having the `put` function have these two function heads?

```elixir
  def put(conn, nil, data, init_options)
    key = Key.generate() |> Key.sign_key(conn)

    put(conn, key, data, init_options)
  end

  def put(_conn, key, data, init_options) do
    set_key_with_retries(key, Value.serialize(data), session_expiration(init_options), 1)
  end
```

File: test/redbird_test.exs [hunk 3]
```diff
@@ -57,6 +57,29 @@ defmodule RedbirdTest do
       assert get_session(conn, :foo) == "bar"
     end
 
+    test "it reuses the session key between requests" do
```
Comment (germsvel) [inline test/redbird_test.exs:60]: Thank you for adding this test! 🎉  It was helpful for me to be able to run it locally to understand the failure you're talking about. 👍

**Review by germsvel — COMMENTED** [review]
@foxbenjaminfox thank you so much for opening this PR. I had a question about whether or not we should be verifying the key when trying to `put` it. Maybe we do? I'm honestly not sure. So I'd love to hear your thoughts on it.

File: lib/redbird/plug/session/redis.ex (hunk 2 above)
Comment (foxbenjaminfox) [reply lib/redbird/plug/session/redis.ex:35]: The `key` that Plug gives us is user-controlled—it's whatever the user sent in the cookie (or `nil`, if the user didn't send the cookie.) This of course is why the signature needs to be verified as part of `get`—but I'm pretty sure the same reasoning applies to `put`. If we don't verify the signature, it seems to me that it would allow the user to set arbitrary keys in redis, which would be a pretty bad security flaw in its own right.

> At the very least, I don't think we should be returning the key in the else clause below since it hasn't passed the Key.accessible? check. In that case, we're not actually setting the key, but by returning it, we're incorrectly telling Plug that we've storing the session.

We have only two options here: we can either return some key, or we can throw an exception. `plug` [puts](https://github.com/elixir-plug/plug/blob/39e1cede12a55496af892bac0c0d7027f5be87ba/lib/plug/session.ex#L96-L97) the value we return directly into the cookie—we don't exactly have a clean way to signal an error.

But we could raise an exception, if you think that's better. Failing silently on a non-valid key is what we do in [`delete/3`](https://github.com/importfeed/redbird/blob/5cabf061d1911461f4c0598dcd3d78be1fc39465/lib/redbird/plug/session/redis.ex#L43-L47)—it would make sense to be consistent about it.

(Actually, for full consistency, if we want an invalid key to throw an error on `put/4` and `delete/3`, wouldn't it make sense to throw an error also on `get/3`, instead of just returning `nil` as if the key was valid but the value was unset?)

Come to think of it, another option would be to treat an invalid (badly-signed, or unsigned) key exactly like how we treat `nil`, and generate a new key to use. That's probably the best solution here.

Comment (germsvel) [reply lib/redbird/plug/session/redis.ex:35]: @foxbenjaminfox thanks so much for all that information. 

>  I'm pretty sure the same reasoning applies to put. If we don't verify the signature, it seems to me that it would allow the user to set arbitrary keys in redis, which would be a pretty bad security flaw in its own right.

That makes sense. So let's keep the verification step in `put` just like you have it now. 

> Come to think of it, another option would be to treat an invalid (badly-signed, or unsigned) key exactly like how we treat nil, and generate a new key to use. That's probably the best solution here.

I like this idea. I don't know if that's somewhat unexpected from the end-user's perspective, but on the other hand, the end user should only care that they get a key back where we've stored the session. So I think that idea sounds great. 

So let's go ahead and treat an invalid key exactly like we treat `nil`. Once that change is in place, I think this is good to go.

Comment (foxbenjaminfox) [reply lib/redbird/plug/session/redis.ex]: > So let's go ahead and treat an invalid key exactly like we treat nil. Once that change is in place, I think this is good to go.

6cc04fd75f270f does this by having `put/4` call itself with `nil` as the second argument if it can't validate the key.

**Comment by foxbenjaminfox** [comment]
799ab47a473a32813d fixes a related issue that I ran into while testing this issue, but it arguably could go into a PR of its own. If you'd prefer that I split it out into a separate PR I will.

`Plug.Crypto.verify/4` defaults to an expiration time of 1 day. This means that signatures were considered invalid after 24 hours, ignoring redbird's `:expiration_in_seconds` option. (Or rather, the redis key would expire in `expiration_in_seconds` seconds, and the signature would expire in 24 hours, and if either expires the session would be considered invalid.)

The fix is simple: I made sure that the `:expiration_in_seconds` option is passed through to `Plug.Crypto.verify/4`'s `:max_age` option, so that the redis expiration and the signature expiration are both the same.

**Review by germsvel — APPROVED** [review]
This looks great. 🎉  Thanks so much for all the work @foxbenjaminfox.  And thank you for offering to split the `max_age` work into a separate one, but I think it's fine for all of it to go in a single PR.

**Comment by germsvel** [comment]
Just pushed this in `v0.7.1`. Thanks for reporting this issue and fixing it!
